File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int pivot=-1;
        int pindex=0;
        int temp=0;
        int tindex=0;
        for(int i=0;i<nums.size()-1;i++){
            if(nums[i]<nums[i+1]){
                pivot=nums[i];
                pindex=i;
            }
        }
        for(int i=nums.size()-1;i>pindex;i--){
            if(nums[i]>nums[pindex]){
                temp=nums[i];
                tindex=i;
                break;
            }
        }
        if(pivot==-1){
            reverse(nums.begin(), nums.end());
            return;
        }
        swap(nums[pindex],nums[tindex]);
        reverse(nums.begin() + pindex+1, nums.end());
    }
};

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int right=nums.length-1;
        int left=-1;
        while(right>0){
            if(nums[right]>nums[right-1]){
                left=right-1;
                break;
            }
            right--;
        }
        if (left == -1) {
            reverse(nums, 0);
            return;
        }
        right=nums.length-1;
        while(right>left){
            if(nums[right]>nums[left]){
                swap(nums, left , right);
                break;
            }
            right--;
        }

        reverse(nums,left+1);
        
    }

    public void swap(int nums[], int left, int right){
        int temp = nums[left];
        nums[left]= nums[right];
        nums[right]=temp;
    }

    public void reverse(int nums[], int left){
        int right=nums.length-1;
        while(left<right){
            int temp = nums[left];
            nums[left] = nums[right];
            nums[right] = temp;
            left++;
            right--;
        }
    }
}